pallene/checker.lua

```diff
@@ -512,6 +575,29 @@ function Checker:check_stat(stat)
         self:expand_function_returns(stat.exps)
 
         for i = 1, #stat.vars do
+            if stat.exps[i]._tag == "ast.Exp.Var" then
+                if stat.exps[i].var._tag == "ast.Var.Dot" then
+                    local top_var = stat.exps[i].var.exp.var
+                    local mod_cname = self.symbol_table:find_symbol(top_var.name)
+                    if mod_cname and mod_cname._tag == "checker.Name.Module" then
+                        local field_name = stat.exps[i].var.name
+                        if not mod_cname.fields[field_name] then
+                            scope_error(stat.exps[i].loc,
+                              "variable '%s.%s' is not declared", top_var.name, field_name)
+                        end
+                    end
+                end
+            end
```

Comment:
Wen we type check an expression we convert qualified identifiers such as `io.write` from an `ast.Var.Dot` to an `ast.Var.Name`. We might want to do the same with module globals.

Reply:
I may be spewing nonsense, because I don't know the code as much as you do, but after the `checker` phase, I believe the module name becomes irrelevant. I could transform it to an `ast.Var.Name`, excluding the module name. We could differentiate it from other `ast.Var.Name` variables by checking that its associated `cname` has the `checker.Mod_field.Var` tag.

Reply:
> We could differentiate it from other ast.Var.Name variables by checking that its associated cname  tag.

Yes, this is the idea. Except that the tag we will want to check will probably be one of the cname tags. That is, `checker.Name.Local`, `checker.Name.Global`, etc.

The `_name` annotation is a `checker.Name` describing what the name means in the current scope. It should either directly contain the relevant information or point to somewhere that does (like the `decl`).